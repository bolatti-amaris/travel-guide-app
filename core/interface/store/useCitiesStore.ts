import { create } from "zustand";
import { devtools, persist, createJSONStorage } from "zustand/middleware";
import AsyncStorage from "@react-native-async-storage/async-storage";

import { City } from "@/core/domain/entities/City";
import { GetAllCities } from "@/core/domain/usecases/GetAllCities";
import { CityRepositoryImpl } from "@/core/infrastructure/repositories/CityRepositoryImpl";

interface CitiesState {
  cities: City[];
  selectedCity: City | null;
  isLoading: boolean;
  error: string | null;
  fetchCities: () => Promise<void>;
  setSelectedCity: (city: City) => void;
  reset: () => void;
}

export const useCitiesStore = create<CitiesState>()(
  devtools(
    persist(
      (set, get) => {
        const repository = new CityRepositoryImpl();
        const getAllCities = new GetAllCities(repository);

        return {
          cities: [],
          selectedCity: null,
          isLoading: false,
          error: null,

          fetchCities: async () => {
            set({ isLoading: true, error: null });

            try {
              const cities = await getAllCities.execute();
              set({ cities, isLoading: false });
              await AsyncStorage.setItem("cities", JSON.stringify(cities));
            } catch (err: any) {
              let message = "Unexpected error";
              if (err.message === "No internet connection") {
                message =
                  "No internet connection. Please check your network and try again.";
              } else if (err.message.toLowerCase().includes("timeout")) {
                message = "The request timed out. Please try again later.";
              } else if (err.message.toLowerCase().includes("graphql")) {
                message =
                  "There was a problem retrieving city data from the server.";
              } else if (err.message) {
                message = err.message;
              }

              const storedCities = await AsyncStorage.getItem("cities");
              if (storedCities) {
                set({ cities: JSON.parse(storedCities), isLoading: false });
              }

              set({
                error: message,
                isLoading: false,
              });
            }
          },

          setSelectedCity: (city) => {
            set({ selectedCity: city });
          },

          reset: () => {
            set(
              {
                cities: [],
                selectedCity: null,
                error: null,
                isLoading: false,
              },
              false,
              "cities/reset",
            );
          },
        };
      },
      {
        name: "cities-storage",
        storage: createJSONStorage(() => AsyncStorage),
      },
    ),
    { name: "CitiesStore" },
  ),
);
