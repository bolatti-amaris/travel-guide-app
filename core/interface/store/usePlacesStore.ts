import { createPlacesStore } from './createPlacesStore';
import { PlaceRepositoryImpl } from '@/core/infrastructure/repositories/PlaceRepositoryImpl';

export const usePlacesStore = createPlacesStore(new PlaceRepositoryImpl());
