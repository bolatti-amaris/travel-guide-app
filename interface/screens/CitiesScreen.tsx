import { useCities } from '@/interface/hooks/useCities';
import CityList from '@/interface/components/CityList';
import React from 'react';
import { View, Text } from 'react-native';
import Loader from '../shared/Loader';
import Retry from '../shared/Retry';

const CitiesScreen = () => {
    const { cities, isLoading, error, fetchCities } = useCities();

    const showEmptyState = !isLoading && !error && cities.length === 0;

  return (
    <View className="flex-1 bg-gray-100 px-4 py-6">
        {isLoading && (
        <Loader message="Loading cities..." />
        )}

        {error && (
        <Retry
        message={error}
        onRetry={fetchCities}
        />
        )}

        {showEmptyState && (
        <Retry
            message="No cities available"
            onRetry={fetchCities}
        />
        )}

        {!isLoading && !error && cities.length > 0 && (
        <CityList cities={cities} />
        )}
    </View>
  );
};

export default CitiesScreen;