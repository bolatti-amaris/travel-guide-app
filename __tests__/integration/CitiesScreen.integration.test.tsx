import React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import CitiesScreen from '@/interface/screens/CitiesScreen';
import { useCitiesStore } from '@/core/interface/store/useCitiesStore';

jest.mock('expo-router', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
    setOptions: jest.fn(),
  }),
  useRouter: () => ({ push: jest.fn() }),
  useSegments: () => [],
}));

jest.mock('@/core/interface/store/useCitiesStore', () => {
  const {
    createCitiesStore,
  } = require('@/core/interface/store/createCitiesStore');
  const mockRepo = {
    getAll: async () => [
      {
        id: 1,
        key: 'barcelona',
        name: 'Barcelona',
        nativeName: 'Barcelona',
        currency: 'EUR',
        language: 'Catalan',
      },
      {
        id: 2,
        key: 'madrid',
        name: 'Madrid',
        nativeName: 'Madrid',
        currency: 'EUR',
        language: 'Spanish',
      },
    ],
  };
  return {
    useCitiesStore: createCitiesStore(mockRepo),
  };
});

describe('CitiesScreen (mocked repository)', () => {
  // Step 1: Reset useCitiesStore store to ensure clean state before each test
  beforeEach(() => {
    useCitiesStore.getState().reset();
  });

  it('renders cities and handles selection', async () => {
    // Step 2: Render the screen wrapped in a NavigationContainer
    render(
      <NavigationContainer>
        <CitiesScreen />
      </NavigationContainer>,
    );

    // Step 3: Wait until the first CityCard is rendered (mocked data)
    await waitFor(() => {
      expect(screen.getByTestId('CityCard-barcelona')).toBeTruthy();
    });

    // Step 4: Assert that both mock cities are present in the list
    expect(screen.getByTestId('CityCard-barcelona')).toBeTruthy();
    expect(screen.getByTestId('CityCard-madrid')).toBeTruthy();

    // Step 5: Simulate press on the first city card
    fireEvent.press(screen.getByTestId('CityCard-barcelona'));

    // Step 6: Validate that selectedCity in useCitiesStore matches the selected item
    const selected = useCitiesStore.getState().selectedCity;
    expect(selected?.key).toBe('barcelona');

    // Step 7: Snapshot the rendered output for consistency checks
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
