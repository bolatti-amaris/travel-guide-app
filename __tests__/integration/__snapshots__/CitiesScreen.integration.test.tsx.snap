// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesScreen (mocked repository) renders cities and handles selection 1`] = `
<View
  className="flex-1 px-4 bg-background dark:bg-background-dark"
  testID="cities-screen"
>
  <RCTScrollView
    data={
      [
        {
          "currency": "EUR",
          "id": 1,
          "key": "barcelona",
          "language": "Catalan",
          "name": "Barcelona",
          "nativeName": "Barcelona",
        },
        {
          "currency": "EUR",
          "id": 2,
          "key": "madrid",
          "language": "Spanish",
          "name": "Madrid",
          "nativeName": "Madrid",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          className="bg-card-background dark:bg-card-background-dark p-4 rounded-sm mb-3"
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="CityCard-barcelona"
        >
          <View
            className="flex-row justify-between items-center"
          >
            <View>
              <View
                className="flex-row flex-wrap items-baseline mb-1"
                testID="city-names"
              >
                <Text
                  className="text-2xl font-work-bold text-text-header dark:text-text-header-dark mr-2"
                  testID="city-name"
                >
                  Barcelona
                </Text>
                <Text
                  className="text-xl font-work-medium text-text dark:text-text-dark italic"
                  testID="city-native-name"
                >
                  (
                  Barcelona
                  )
                </Text>
              </View>
              <View
                className="flex-row flex-wrap"
                testID="city-details"
              >
                <Text
                  className="text-base text-text dark:text-text-dark font-work-light mr-2"
                  testID="city-currency"
                >
                  Currency: 
                  EUR
                </Text>
                <Text
                  className="text-base text-text dark:text-text-dark font-work-light"
                  testID="city-language"
                >
                  Language: 
                  Catalan
                </Text>
              </View>
            </View>
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={24}
              bbWidth={24}
              fill="none"
              focusable={false}
              height={24}
              meetOrSlice={0}
              minX={0}
              minY={0}
              stroke="#262626"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  {
                    "flex": 0,
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
              vbHeight={24}
              vbWidth={24}
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <RNSVGGroup
                fill={null}
                propList={
                  [
                    "fill",
                    "stroke",
                    "strokeWidth",
                    "strokeLinecap",
                    "strokeLinejoin",
                  ]
                }
                stroke={
                  {
                    "payload": 4280690214,
                    "type": 0,
                  }
                }
                strokeLinecap={1}
                strokeLinejoin={1}
                strokeWidth={2}
              >
                <RNSVGPath
                  d="m9 18 6-6-6-6"
                  fill={null}
                  propList={
                    [
                      "fill",
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={
                    {
                      "payload": 4280690214,
                      "type": 0,
                    }
                  }
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth={2}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          className="bg-card-background dark:bg-card-background-dark p-4 rounded-sm mb-3"
          collapsable={false}
          focusable={true}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="CityCard-madrid"
        >
          <View
            className="flex-row justify-between items-center"
          >
            <View>
              <View
                className="flex-row flex-wrap items-baseline mb-1"
                testID="city-names"
              >
                <Text
                  className="text-2xl font-work-bold text-text-header dark:text-text-header-dark mr-2"
                  testID="city-name"
                >
                  Madrid
                </Text>
                <Text
                  className="text-xl font-work-medium text-text dark:text-text-dark italic"
                  testID="city-native-name"
                >
                  (
                  Madrid
                  )
                </Text>
              </View>
              <View
                className="flex-row flex-wrap"
                testID="city-details"
              >
                <Text
                  className="text-base text-text dark:text-text-dark font-work-light mr-2"
                  testID="city-currency"
                >
                  Currency: 
                  EUR
                </Text>
                <Text
                  className="text-base text-text dark:text-text-dark font-work-light"
                  testID="city-language"
                >
                  Language: 
                  Spanish
                </Text>
              </View>
            </View>
            <RNSVGSvgView
              align="xMidYMid"
              bbHeight={24}
              bbWidth={24}
              fill="none"
              focusable={false}
              height={24}
              meetOrSlice={0}
              minX={0}
              minY={0}
              stroke="#262626"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth={2}
              style={
                [
                  {
                    "backgroundColor": "transparent",
                    "borderWidth": 0,
                  },
                  {
                    "flex": 0,
                    "height": 24,
                    "width": 24,
                  },
                ]
              }
              vbHeight={24}
              vbWidth={24}
              width={24}
              xmlns="http://www.w3.org/2000/svg"
            >
              <RNSVGGroup
                fill={null}
                propList={
                  [
                    "fill",
                    "stroke",
                    "strokeWidth",
                    "strokeLinecap",
                    "strokeLinejoin",
                  ]
                }
                stroke={
                  {
                    "payload": 4280690214,
                    "type": 0,
                  }
                }
                strokeLinecap={1}
                strokeLinejoin={1}
                strokeWidth={2}
              >
                <RNSVGPath
                  d="m9 18 6-6-6-6"
                  fill={null}
                  propList={
                    [
                      "fill",
                      "stroke",
                      "strokeWidth",
                      "strokeLinecap",
                      "strokeLinejoin",
                    ]
                  }
                  stroke={
                    {
                      "payload": 4280690214,
                      "type": 0,
                    }
                  }
                  strokeLinecap={1}
                  strokeLinejoin={1}
                  strokeWidth={2}
                />
              </RNSVGGroup>
            </RNSVGSvgView>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
